File: leet/0212_word_search_2/0212_word_search_2.py
# LeetCode 212. Word Search II
# https://leetcode.com/problems/word-search-ii/
# O(k*m*n) time O(k*w) space, k average word length, w words
class Trie:
    def __init__(self):
        self.word=None
        self.children=defaultdict(Trie)
    def insert(self,word):
        node = self
        for c in word:
            node = node.children[c]    
        node.word = word
        
    def delete (self, node, word, i):
        if i == len(word):
            if node.word == word:
                node.word = None
            if not node.children:
                return True
            return False
        if word[i] not in node.children:
            return False
        if self.delete(node.children[word[i]], word, i + 1):
            del node.children[word[i]]
            if not node.children:
                return True
        return False
        
class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        ans = set()
        def backtrack(i,j,node):
            c=board[i][j]
            if c not in node.children:
                return
            if node.children[c].word != None:
                ans.add(node.children[c].word)
                root.delete(root, node.children[c].word, 0)
            
            board[i][j] = '*'
            if i>0:
                backtrack(i-1,j,node.children[c])
            if j>0:
                backtrack(i,j-1,node.children[c])
            if i<len(board)-1:
                backtrack(i+1,j,node.children[c])
            if j<len(board[i])-1:
                backtrack(i,j+1,node.children[c])
            board[i][j] = c
    
        root = Trie()
        for word in words:
            root.insert(word)
        for i in range(len(board)):
            for j in range(len(board[i])):
                backtrack(i,j,root)
        return list(ans)

File: leet/0212_word_search_2/0212_word_search_2.ts
// LeetCode 212. Word Search II
// https://leetcode.com/problems/word-search-ii/
// O(m*n*3^l) time, O(w*l) space
class Trie {
    children: {[key: string]: Trie};
    isWord: boolean;
    numWords: number;

    constructor() {
        this.children = {};
        this.isWord = false;
        this.numWords = 0;
    }

    insert(word: string): void {
        let node: Trie = this;
        node.numWords += 1;
        for (let c of word) {
            node = c in node.children ? node.children[c]
                : node.children[c] = new Trie();
            node.numWords += 1;
        }
        node.isWord = true;
    }

    delete(word: string) {
        let node: Trie = this;
        node.numWords -= 1;
        for (let c of word) {
            node = node.children[c];
            node.numWords -= 1;
        }
        node.isWord = false;
    }

}

function findWords(board: string[][], words: string[]): string[] {
    const trie = new Trie();
    const solutions = [];
    const backtrack = (r, c, node, maybeWord) => {
        const char = board[r][c], nextMaybeWord = maybeWord + char,
              nextNode = node.children[char];
        board[r][c] = '#';
        if (nextNode) {
            let numWords = nextNode.numWords, isWord = nextNode.isWord;
            if (isWord) {
                solutions.push(nextMaybeWord);
                numWords--;
                trie.delete(nextMaybeWord);
            }
            if (numWords > 0) {
                for (let [dr, dc] of [[0, 1], [0, -1], [1, 0], [-1, 0]]) {
                    const nextR = r + dr, nextC = c + dc;
                    if (nextR < 0 || nextR >= board.length || nextC < 0 ||
                        nextC >= board[0].length || board[nextR][nextC] == '#') {
                        continue;
                    }
                    backtrack(nextR, nextC, nextNode, nextMaybeWord);
                }
            }
        }
        board[r][c] = char;
    }
    for (let word of words) {
        trie.insert(word);
    }
    for (let r = 0; r < board.length; r++) {
        for (let c = 0; c < board[0].length; c++) {
            backtrack(r, c, trie, '');
        }
    }
    return solutions;
};
